### PR TITLE
Modernize CMake

### PR DESCRIPTION
**Pull request by SimeonEhrig**
The CMake is completely modernized. The main targets are 
- make cupla installable and findable via cmake `find_package()`
- make the usage of `add_subdirectory()` more common 
- remove legacy cmake code

Address issue #158 

# TODO

- [x] setup initial CMakeLists.txt
- [x] remove legacy CMake code 
- [x] overwork usage via `add_subdirectory`
- [x] allow to install cupla and use `find_package()`
- [x] add CI integration tests
- [x] enable examples
- [x] writing documentation!
- [x] create cupla release branch before this PR is merged

### CODE REVIEW

File: CMakeLists.txt
Comment (j-stephan) [inline CMakeLists.txt:2]: Needs an updated year here.

File: CMakeLists.txt
Comment (j-stephan) [inline CMakeLists.txt:66]: This should be `internal` (and `external`).

File: CMakeLists.txt
Comment (j-stephan) [inline CMakeLists.txt:61]: What happens if someone tries this with alpaka 0.5?

**Review by j-stephan — CHANGES_REQUESTED** [review]

File: CMakeLists.txt
Comment (SimeonEhrig) [reply CMakeLists.txt:61]: Should throw an an error at cmake configure time, because cupla 0.3 is not compatible with the old alpaka interface.

Comment (SimeonEhrig) [reply CMakeLists.txt:61]: By the way. I also want to go to CMake 3.19. Then we have support for version ranges. I think, this is quit useful for alpaka. But I'm not sure, if alpaka also need to use CMake 3.19. I have to test this.

Comment (j-stephan) [reply CMakeLists.txt:61]: This should work without alpaka having to change. Otherwise you'd have to manually change the CMake version of every third party library you want to use.

Comment (SimeonEhrig) [reply CMakeLists.txt:61]: I also thought it. My first try failed. I could use an unsupported alpaka version, but it is really possible, that I did a mistake.

**Comment by SimeonEhrig** [comment]
@psychocoderHPC Ready for review

**Comment by SimeonEhrig** [comment]
From the review in our developer meeting today:
- [ ] force to build libcupla as archive file -> makes no sense to use shared lib, because libcupla is strongly depend of the build parameter of the project and avoids `LD_LIBRARY_PATH` problem
- [ ] downgrade to CMake 3.18 -> same version like alpaka - we don't use any advantage of CMake 3.19

**Comment by psychocoderHPC** [comment]
@SimeonEhrig Do you tested this PR with nvcc as CUDA compiler too? I working on moving alpaka 0.7.0 to cupla and integrated your changes. When try to use cmake with CUDA I get 

```
cmake ../cupla -DCMAKE_INSTALL_PREFIX=$HOME/lib/cupla -Dcupla_BUILD_EXAMPLES=ON -DALPAKA_ACC_GPU_CUDA_ENABLE=ON
...
List of compiler flags added by alpaka
device compiler:
    $<$<COMPILE_LANGUAGE:CUDA>:--expt-relaxed-constexpr>;$<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe=--display_error_number>;$<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe=--diag_suppress=esa_on_defaulted_function_ignored>

CMake Error at cmake/cuplaTargetHelper.cmake:38 (target_include_directories):
  Cannot specify include directories for target "cupla" which is not built by
  this project.
Call Stack (most recent call first):
  CMakeLists.txt:101 (createCuplaTarget)


-- Configuring incomplete, errors occurred!
See also "/home/rwidera/workspace/buildCupla/CMakeFiles/CMakeOutput.log".

```

File: cmake/cuplaTargetHelper.cmake
Comment (psychocoderHPC) [inline cmake/cuplaTargetHelper.cmake:60]: This must be `PRIVATE` we do not enforce the user to use any compiler flags if not needed.
The same must be changed in the other if branches too.

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

**Comment by psychocoderHPC** [comment]
> @SimeonEhrig Do you tested this PR with nvcc as CUDA compiler too? I working on moving alpaka 0.7.0 to cupla and integrated your changes. When try to use cmake with CUDA I get
> 
> ```
> cmake ../cupla -DCMAKE_INSTALL_PREFIX=$HOME/lib/cupla -Dcupla_BUILD_EXAMPLES=ON -DALPAKA_ACC_GPU_CUDA_ENABLE=ON
> ...
> List of compiler flags added by alpaka
> device compiler:
>     $<$<COMPILE_LANGUAGE:CUDA>:--expt-relaxed-constexpr>;$<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe=--display_error_number>;$<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe=--diag_suppress=esa_on_defaulted_function_ignored>
> 
> -- Configuring done
> CMake Error in CMakeLists.txt:
>   No known features for CUDA compiler
> 
>   ""
> 
>   version .
> ```

This error happens only with `alpaka 0.7.0` so ignore it for now.

**Comment by SimeonEhrig** [comment]
> > @SimeonEhrig Do you tested this PR with nvcc as CUDA compiler too? I working on moving alpaka 0.7.0 to cupla and integrated your changes. When try to use cmake with CUDA I get
> > ```
> > cmake ../cupla -DCMAKE_INSTALL_PREFIX=$HOME/lib/cupla -Dcupla_BUILD_EXAMPLES=ON -DALPAKA_ACC_GPU_CUDA_ENABLE=ON
> > ...
> > List of compiler flags added by alpaka
> > device compiler:
> >     $<$<COMPILE_LANGUAGE:CUDA>:--expt-relaxed-constexpr>;$<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe=--display_error_number>;$<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe=--diag_suppress=esa_on_defaulted_function_ignored>
> > 
> > -- Configuring done
> > CMake Error in CMakeLists.txt:
> >   No known features for CUDA compiler
> > 
> >   ""
> > 
> >   version .
> > ```
> 
> This error happens only with `alpaka 0.7.0` so ignore it for now.

We can't ignore `alpaka 0.7.0`. This PR should become part of `cupla 0.4.0` which bases on `alpaka 0.7.0`.

**Comment by psychocoderHPC** [comment]
> > > @SimeonEhrig Do you tested this PR with nvcc as CUDA compiler too? I working on moving alpaka 0.7.0 to cupla and integrated your changes. When try to use cmake with CUDA I get
> > > ```
> > > cmake ../cupla -DCMAKE_INSTALL_PREFIX=$HOME/lib/cupla -Dcupla_BUILD_EXAMPLES=ON -DALPAKA_ACC_GPU_CUDA_ENABLE=ON
> > > ...
> > > List of compiler flags added by alpaka
> > > device compiler:
> > >     $<$<COMPILE_LANGUAGE:CUDA>:--expt-relaxed-constexpr>;$<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe=--display_error_number>;$<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe=--diag_suppress=esa_on_defaulted_function_ignored>
> > > 
> > > -- Configuring done
> > > CMake Error in CMakeLists.txt:
> > >   No known features for CUDA compiler
> > > 
> > >   ""
> > > 
> > >   version .
> > > ```
> > 
> > 
> > This error happens only with `alpaka 0.7.0` so ignore it for now.
> 
> We can't ignore `alpaka 0.7.0`. This PR should become part of `cupla 0.4.0` which bases on `alpaka 0.7.0`.

I played a little bit more with the error. If I install alpaka and use `-Dcupla_ALPAKA_PROVIDER=external` all works fine.
It could be that the issue will be solved with https://github.com/Kitware/CMake/commit/a3cafa4237930d1ec1bcbdddc8a9c592e32458ca but this is currently in no release available.

**Comment by psychocoderHPC** [comment]
I tested the latest master branch from CMake, the issue still exists, therefore something else is wrong :-(

**Review by psychocoderHPC — APPROVED** [review]

**Comment by psychocoderHPC** [comment]
We will not merge this PR directly, if #213 is merged this PR will be merged automatically.

**Comment by fwyzard** [comment]
>  force to build libcupla as archive file -> makes no sense to use shared lib, because libcupla is strongly depend of the build parameter of the project and avoids `LD_LIBRARY_PATH` problem

FWIW, in CMSSW we have been building cupla as a shared library, even though we are not using it in production yet.
The rationale is that it would be linked to from many different shared libraries, so using a static archive would duplicate it into each shared library that uses it.

**Comment by SimeonEhrig** [comment]
> > force to build libcupla as archive file -> makes no sense to use shared lib, because libcupla is strongly depend of the build parameter of the project and avoids `LD_LIBRARY_PATH` problem
> 
> FWIW, in CMSSW we have been building cupla as a shared library, even though we are not using it in production yet.
> The rationale is that it would be linked to from many different shared libraries, so using a static archive would duplicate it into each shared library that uses it.

Good point. I opened an issue (#218 ) to address your comment.

**Pull request merged**
